### PR TITLE
Add date range component

### PR DESCRIPTION
**Pull request by StevePole**

### CODE REVIEW

File: angular/services/tw-date-service/tw-date-service.service.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
+(function(angular) {
+	'use strict';
+
+	angular
+		.module('tw.form-components')
+		.service('TwDateService', TwDateService);
+
+	function TwDateService() {
+		var DEFAULT_MONTHS_EN = [
```
Comment (skywritergr) [inline angular/services/tw-date-service/tw-date-service.service.js]: Do we care about translations here?

File: angular/services/tw-date-service/tw-date-service.service.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+			date.setUTCHours(0);
+			date.setUTCMinutes(0);
+			date.setUTCSeconds(0);
+			date.setUTCMilliseconds(0);
+			return date;
+		};
+
+		this.getDayNamesForLocale = function(locale, format) {
+			if (!format) {
+				format = 'long';  // or 'narrow' or 'short'
```
Comment (skywritergr) [inline angular/services/tw-date-service/tw-date-service.service.js]: What do you mean with that comment here?

File: angular/services/tw-date-service/tw-date-service.service.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+				// Day in middle of month avoids timezone issues
+				date = this.getUTCDateFromParts(2001, 0, i); // First day of millenium was monday
+				var dayName = date.toLocaleDateString(locale, {weekday: format});
+				dayName = dayName[0].toUpperCase() + dayName.substring(1);
+				days.push(dayName);
+			}
+			return days;
+		};
+
+		this.getMonthNamesForLocale = function(locale, format) {
```
Comment (skywritergr) [inline angular/services/tw-date-service/tw-date-service.service.js]: The code between here and the function above shares a lot of similarities. Can we extract some of this code in a function that you can run it in both places, or merge the code somehow?

File: index.js [hunk 4]
```diff
@@ -197,6 +214,10 @@ angular.module('my-app', ['tw.styleguide-components'])
 			{value: 'xs,sm', label: 'xs and sm grid'},
 			{value: 'md,lg,xl', label: 'md, lg and xl grid'}
 		];
+
+		$ctrl.log = function(message) {
```
Comment (skywritergr) [inline index.js]: Why is this necessary?

Comment (StevePole) [inline index.js]: This is just used by the demo page, it allows us to 
```
ng-change="$ctrl.log('change')"
ng-blur="$ctrl.log('blur')"
```
etc.  
Those values must be angular expressions, which precludes using console.log

File: angular/components/tw-requirements-form/tw-requirements-form.directive.js [hunk 5]
```diff
@@ -155,7 +155,7 @@
 		function prepRegExp(field) {
 			if (field.validationRegexp) {
 				try {
-					field.validationRegexp = new RegExp(ield.validationRegexp);
+					field.validationRegexp = new RegExp(field.validationRegexp);
```
Comment (StevePole) [inline angular/components/tw-requirements-form/tw-requirements-form.directive.js]: You would expect this regex to come from the API, so I can't say what it would do at this point.

File: angular/services/tw-date-service/tw-date-service.service.js (hunk 1 above)
Comment (StevePole) [inline angular/services/tw-date-service/tw-date-service.service.js]: This is a fallback, it most browsers we can get the localised month name from the Date object, but in case the locale isn't available, or the browser does not support localised dates, then we can fall back to these.

File: angular/services/tw-date-service/tw-date-service.service.js (hunk 2 above)
Comment (StevePole) [inline angular/services/tw-date-service/tw-date-service.service.js]: These are the values supported by the Date object's formatting function for day names.  I think useful to be able to see, but I will change it to be a bit clearer.

File: angular/services/tw-date-service/tw-date-service.service.js (hunk 3 above)
Comment (StevePole) [inline angular/services/tw-date-service/tw-date-service.service.js]: Yeah, good point, I'l see what I can do.

File: angular/components/tw-requirements-form/tw-requirements-form.directive.js (hunk 5 above)
Comment (skywritergr) [inline angular/components/tw-requirements-form/tw-requirements-form.directive.js]: Ok, got it.

File: angular/services/tw-date-service/tw-date-service.service.js (hunk 3 above)
Comment (StevePole) [inline angular/services/tw-date-service/tw-date-service.service.js]: Made some changes, hard to do more as the key remaining lines differ in small ways.

File: angular/components/tw-date-lookup/tw-date-lookup.directive.js [hunk 6]
```diff
@@ -184,9 +184,14 @@
...
 
+			ngModelCtrl.$formatters.push(function(newDate) {
+				updateCalendarView(newDate);
+				return newDate;
+			});
+
 			$element.find('.btn, .dropdown-menu').on('focusout', function() {
 				$timeout(function() {
-					// If button isn't focused and dropdown not open, blur
+					// If button isn't focused and dropdown not , blur
```
Comment (skywritergr) [inline angular/components/tw-date-lookup/tw-date-lookup.directive.js]: not what?

File: angular/components/tw-date-lookup/tw-date-lookup.directive.js [hunk 7]
```diff
@@ -197,22 +202,24 @@
...
-			$ctrl.resetView($ctrl.ngModel);
 			setLocale($ctrl.locale);
-			 // Don't update viewModel on this call
-			updateModelDate($ctrl.ngModel, true);
-			updateMinDate($ctrl.ngMin);
-			updateMaxDate($ctrl.ngMax);
-			ngModelCtrl.$validate();
 
-			$ctrl.weeks = getTableStructure();
+			updateMinDateView($ctrl.ngMin);
```
Comment (skywritergr) [inline angular/components/tw-date-lookup/tw-date-lookup.directive.js]: nitpicking here, but why ngMin and not just min?

Comment (StevePole) [inline angular/components/tw-date-lookup/tw-date-lookup.directive.js]: Because we're trying to follow the interface used by native controls in conjunction with Angular.  Normally you would do:
```
<input type="date" ng-modal="..." ng-min="..." />
```
With this you do:
```
<tw-date-lookup ng-model="..." ng-min="..." />
```
Makes migration easier.

File: dist/js/form-components.js [hunk 8]
```diff
@@ -317,40 +317,48 @@ angular.module("tw.form-components", []);
...
+            }), setLocale($ctrl.locale), updateMinDateView($ctrl.ngMin), updateMaxDateView($ctrl.ngMax);
         }
         function resetFocus() {
             $element.find("button").focus();
         }
         function addValidators() {
             ngModelCtrl.$validators.min = function(modelValue, viewValue) {
                 var value = modelValue || viewValue;
-                return !(value && value < $ctrl.ngMin) || (unsetModel(), !1);
+                return !(value && value < $ctrl.ngMin) || ($element.parents(".form-group").addClass("has-error"),
```
Comment (skywritergr) [inline dist/js/form-components.js]: This thing here looks more complicated than it should. Is it returning things or is it setting things? It seems like it's doing both. I wouldn't mind if it was a bit more verbose and readable.

File: dist/js/form-components.js [hunk 9]
```diff
@@ -317,40 +317,48 @@ angular.module("tw.form-components", []);
...
         function addValidators() {
             ngModelCtrl.$validators.min = function(modelValue, viewValue) {
                 var value = modelValue || viewValue;
-                return !(value && value < $ctrl.ngMin) || (unsetModel(), !1);
+                return !(value && value < $ctrl.ngMin) || ($element.parents(".form-group").addClass("has-error"), 
+                !1);
             }, ngModelCtrl.$validators.max = function(modelValue, viewValue) {
                 var value = modelValue || viewValue;
-                return !(value && value > $ctrl.ngMax) || (unsetModel(), !1);
+                return !(value && value > $ctrl.ngMax) || ($element.parents(".form-group").addClass("has-error"),
```
Comment (skywritergr) [inline dist/js/form-components.js]: Same like the comment above.

File: dist/js/form-components.js [hunk 10]
```diff
@@ -370,23 +378,21 @@ angular.module("tw.form-components", []);
             $ctrl.daysOfWeek = TwDateService.getDayNamesForLocale($ctrl.locale, "short"), $ctrl.shortDaysOfWeek = TwDateService.getDayNamesForLocale($ctrl.locale, "narrow");
         }
         function moveDateToWithinRange(date, min, max) {
-            return min && min > date ? min : max && max < date ? max : date;
+            return date || (date = TwDateService.getLocaleToday()), min && min > date ? min : max && max < date ? max : date;
```
Comment (skywritergr) [inline dist/js/form-components.js]: That would greatly benefit from parenthesis or/and splitting it a bit.

File: angular/components/tw-date-lookup/tw-date-lookup.directive.js (hunk 6 above)
Comment (StevePole) [inline angular/components/tw-date-lookup/tw-date-lookup.directive.js]: Weird, I'll put that back.

File: dist/js/form-components.js (hunk 8 above)
Comment (StevePole) [inline dist/js/form-components.js]: Just the minifier getting excited, would read the original source code in date-lookup

File: dist/js/form-components.js (hunk 10 above)
Comment (StevePole) [inline dist/js/form-components.js]: This is the minifier getting a bit excited, best to check the source file.
